src/modules/party/party-customer-type.service.ts

```diff
@@ -0,0 +1,43 @@
+import { Injectable } from '@nestjs/common';
+import { MdmService } from '@ukef/modules/mdm/mdm.service';
+import { PinoLogger } from 'nestjs-pino';
+
+import { MdmResourceNotFoundException } from '../mdm/exception/mdm-resource-not-found.exception';
+
+type CustomerType = string | null;
+
+@Injectable()
+export class PartyCustomerTypeService {
+  constructor(private readonly mdmService: MdmService, private readonly logger: PinoLogger) {}
+
+  async getCustomerTypeForPartyFromAlternateIdentifier({
+    alternateIdentifier,
+    fallbackIfNotFound,
+  }: {
+    alternateIdentifier: string;
+    fallbackIfNotFound: CustomerType;
+  }): Promise<CustomerType> {
+    try {
+      const customersWithPartyUrnMatchingAlternateIdentifier = await this.mdmService.findCustomersByPartyUrn(alternateIdentifier);
+
+      if (customersWithPartyUrnMatchingAlternateIdentifier.length === 0) {
+        this.logThatNoCustomersWereFound({ alternateIdentifier, fallbackIfNotFound });
+        return fallbackIfNotFound;
+      }
+
+      const [{ type: typeOfFirstCustomerFound }] = customersWithPartyUrnMatchingAlternateIdentifier;
+      return typeOfFirstCustomerFound;
+    } catch (error) {
+      if (error instanceof MdmResourceNotFoundException) {
+        // TODO APIM-336: Should we check the error message as well to not apply this logic for unexpected 404 errors?
+        this.logThatNoCustomersWereFound({ alternateIdentifier, fallbackIfNotFound });
+        return fallbackIfNotFound;
+      }
+      throw error;
```

Comment:
@avaitonis FYI this is probably the most interesting bit of the existing Mulesoft logic: if there is a 404 error (which could either be that no matching customers were found _or_ an unexpected 404 error e.g. if the MDM base URL is configured incorrectly) then we will create the external rating with a CORPORATE assigned rating code. If there's any other error, then we will return a 500 error and not create the external rating

I'm going to keep it as-is since we want a like-for-like migration, but I will raise a ticket for the backlog that this could be made more sensible (but this would involve changes outside of TFS)

Reply:
Created APIM-493